numpy/linalg/umath_linalg.c.src

```diff
@@ -3213,35 +3251,70 @@ static void
     GELSD_PARAMS_t params;
     int error_occurred = get_fp_invalid_and_clear();
     fortran_int n, m, nrhs;
-    INIT_OUTER_LOOP_6
+    fortran_int excess;
+    fortran_int one = 1;
+
+    INIT_OUTER_LOOP_7
 
     m = (fortran_int)dimensions[0];
     n = (fortran_int)dimensions[1];
     nrhs = (fortran_int)dimensions[2];
+    excess = m - n;
 
     if (init_@lapack_func@(&params, m, n, nrhs)) {
-        LINEARIZE_DATA_t a_in, b_in, x_out, s_out;
+        LINEARIZE_DATA_t a_in, b_in, x_out, s_out, r_out;
 
         init_linearize_data(&a_in, n, m, steps[1], steps[0]);
         init_linearize_data_ex(&b_in, nrhs, m, steps[3], steps[2], fortran_int_max(n, m));
-        init_linearize_data(&x_out, nrhs, fortran_int_max(n, m), steps[5], steps[4]);
-        init_linearize_data(&s_out, 1, fortran_int_min(n, m), 1, steps[6]);
+        init_linearize_data_ex(&x_out, nrhs, n, steps[5], steps[4], fortran_int_max(n, m));
+        init_linearize_data(&r_out, 1, nrhs, 1, steps[6]);
+        init_linearize_data(&s_out, 1, fortran_int_min(n, m), 1, steps[7]);
 
-        BEGIN_OUTER_LOOP_6
+        BEGIN_OUTER_LOOP_7
             int not_ok;
             linearize_@TYPE@_matrix(params.A, args[0], &a_in);
             linearize_@TYPE@_matrix(params.B, args[1], &b_in);
             params.RCOND = args[2];
             not_ok = call_@lapack_func@(&params);
             if (!not_ok) {
                 delinearize_@TYPE@_matrix(args[3], params.B, &x_out);
-                *(npy_int*) args[4] = params.RANK;
-                delinearize_@REALTYPE@_matrix(args[5], params.S, &s_out);
+                *(npy_int*) args[5] = params.RANK;
+                delinearize_@REALTYPE@_matrix(args[6], params.S, &s_out);
+
+                if (excess >= 0 && params.RANK == n) {
```

Comment:
Note that `np.linalg.lstsq` only actually returns residuals if `excess > 0`. I'd consider this a bug, as to me the residuals are well defined as `0` when `m == n`.

Either way, I've deliberately not changed the external interface yet.